File: 프로그래머스/2/42746. 가장 큰 수/가장 큰 수.java
import java.util.*;

class Solution {
    public String solution(int[] numbers) {
        List<String> numbersList = new ArrayList<>();
        for(int num : numbers){
            numbersList.add(String.valueOf(num)); // 체크
        }
        numbersList.sort((o1, o2) -> (o2+o1).compareTo(o1+o2)); // 외워
        StringBuilder sb = new StringBuilder();
        for(String num : numbersList){
            sb.append(num);
        }
        
        String answer = sb.toString();
        return answer.startsWith("0") ? "0" : answer;
    }
}

// import java.util.*;

// class Solution {
//     public String solution(int[] numbers) {
//         String[] nums = new String[numbers.length];
//         for(int i = 0; i < numbers.length; i++){
//             nums[i] = String.valueOf(numbers[i]);
//         }
        
//         Arrays.sort(nums, (o1, o2) -> (o2 + o1).compareTo(o1 + o2));
        
//         if(nums[0].equals("0")) return "0";
        
//         StringBuilder sb = new StringBuilder();
//         for(String num : nums){
//             sb.append(num);
//         }
//         return sb.toString();
//     }
// }

File: 프로그래머스/2/42746. 가장 큰 수/가장 큰 수.kt
import java.util.Comparator

class Solution {
    fun solution(numbers: IntArray): String {
        // 1. 숫자 배열을 문자열로 변환
        val strNumbers = mutableListOf<String>()
        for(num in numbers){
            strNumbers.add(num.toString())
        }

        // 2. 정렬 기준 설정: (b+a) 가 (a+b)보다 크면 b가 앞으로 오도록 설정
        // 3+30 > 30+3 : 이 경우 3 30 순서
        val sortedNumbers = strNumbers.sortedWith(Comparator{
            a, b ->
            (b + a).compareTo(a+b) // 내림차순 정렬
        })

        // 3. 정렬된 것 이어붙여 결과 생성
        val sb = StringBuilder()
        sortedNumbers.forEach{sb.append(it)}
        val result = sb.toString()
        //println(result)
        // 4. "000" 같은 경우 "0"으로 변환
        return if(result.startsWith("0")) "0" else result
    }
}